firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,391 @@
+      }
+    }
+  }
+
+  GIVEN("A CRC element is initalised with valid payload parameter") {
+    constexpr size_t buffer_size = 254UL;
+    using TestCRCElementProps = PF::Protocols::CRCElementProps<buffer_size>;
+    using TestCRCElement = PF::Protocols::CRCElement<TestCRCElementProps::PayloadBuffer>;
+    PF::HAL::SoftCRC32 crc32c{PF::HAL::crc32c_params};
+
+    auto expected_payload = std::string("\x01\x05\x01\x02\x03\x04\x05", 7);
+    auto input_data = PF::Util::make_array<uint8_t>(0x01, 0x05, 0x01, 0x02, 0x03, 0x04, 0x05);
+    TestCRCElementProps::PayloadBuffer input_payload;
+
+    for (auto& data : input_data) {
+      input_payload.push_back(data);
+    }
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("a body with good crc and payload fields is parsed") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x98DBE355;
+
+      auto body = std::string("\x98\xdb\xe3\x55\x01\x05\x01\x02\x03\x04\x05", 11);
+      auto expected_payload = std::string("\x01\x05\x01\x02\x03\x04\x05", 7);
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto parse_status = crc_element.parse(input_buffer);
+      auto crc_check = crc_element.crc();
+      const auto &payload_check = crc_element.payload();
+
+      THEN("the parse status should be ok") {
+        REQUIRE(parse_status == PF::IndexStatus::ok);
+      }
+
+      THEN("crc from the CRCElement is equal to the expected crc for the given payload") {
+        REQUIRE(crc_check == expected_crc);
+      }
+
+      THEN("payload from the CRCElement is equal to the expected payload from the body") {
+        REQUIRE(payload_check == expected_payload);
+      }
+    }
+
+    WHEN("body less than size of 4 bytes is parsed") {
+      auto body = std::string("\x98\xdb\xe3", 3);
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto parse_status = crc_element.parse(input_buffer);
+
+      THEN("the parse status is out of bounds") {
+        REQUIRE(parse_status == PF::IndexStatus::out_of_bounds);
+      }
+    }
+
+    WHEN("body with invalid crc is parsed") {
+      auto body = std::string("\x12\x34\x56\x78\x03\x04\x00\xed\x30\x00", 10);
+      auto expected_payload = std::string("\x03\x04\x00\xed\x30\x00", 6);
+
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x6D551F4C;
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto parse_status = crc_element.parse(input_buffer);
+      auto crc_check = crc_element.crc();
+
+      auto payload_check = crc_element.payload();
+
+      THEN("the parse status is ok") {
+        REQUIRE(parse_status == PF::IndexStatus::ok);
+      }
+
+      THEN("crc from the CRCElement is not equal to the expected crc for the given payload") {
```

Comment:
We should add another assertion: the CRC from the CRCElement is equal to what was encoded in the body. Since what was encoded in the body is hardcoded to be different from what would be computed using the payload, this new assertion can completely replace what we have here.

Reply:
updated.